### ISSUE TITLE
Open for comment: DX Images (Developer Experience)

### ISSUE THREAD

**Issue by KyleGospo** [ublue-os/bazzite#533]
Similarly to Bluefin-DX, we'd like to ship DX images which come preconfigured with common game development and debugging tools. If you're a developer we'd love to hear what you might find useful in an image like Bazzite.

**Comment by EyeCantCU** [comment]
Suggestions:
* godot: https://godotengine.org/
* android-tools: Provides ADB, needed for Android game development

Utilities that would likely be helpful:
* Blender
* GIMP
* InkScape
* Audacity

Things that would be cool, but totally not necessary/probably out of scope:
* Zsh - with oh-my-zsh configured with the half life theme
* chatgpt-shell-cli symlinked to GLaDOS/glados
* Nerd fonts - broader symbol support
* bat, eza, fd - robust alternatives to cat, ls, and find
* virt-manager - for image testing

**Comment by badlydrawnface** [comment]
That would seem interesting! Getting JetBrains IDEs installed through a copr is a bit of a hassle and there doesn't seem to be a way to make JetBrains Toolbox run on a specific Distrobox container (obviously excluding the AUR package), because of how it installs, it can only install on the host, where there won't be any (other than probably Python) SDKs installed.

**Comment by EyeCantCU** [comment]
> That would seem interesting! Getting JetBrains IDEs installed through a copr is a bit of a hassle and there doesn't seem to be a way to make JetBrains Toolbox run on a specific Distrobox container (obviously excluding the AUR package), because of how it installs, it can only install on the host, where there won't be any (other than probably Python) SDKs installed.

The challenge here is, you can't throw all of them into one distrobox image (nor do people need all of them... usually) as the size is too big for GHCR unless you do it after the fact

One solution would be using distrobox assemble for each JetBrains IDE, keeping the environment completely separate from one another, installing everything while setting the image up on the host

Another alternative is one distrobox instance that provides everything needed by each IDE with just scripts that install whatever IDE you need

**Comment by castrojo** [comment]
Jetbrains has it's own "toolbox" that people have been using and so far the feedback has been positive: https://universal-blue.org/images/bluefin/developer-experience/#jetbrains

**Comment by EyeCantCU** [comment]
> Jetbrains has it's own "toolbox" that people have been using and so far the feedback has been positive: https://universal-blue.org/images/bluefin/developer-experience/#jetbrains

Toolbox works great for that but these are all installed as AppImages that rely on the host to provide individual SDKs. Providing a path isolated from the host that's reusable anywhere could be an awesome way of managing that. Brainstorming ways of doing that

**Comment by bi0shacker001** [comment]
As a reminder, distrobox has a --home flag that lets you specify an alternative home directory for the container user, which, coupled with `distrobox-enter -n (containername) -- (executable in relative location)` , may solve the issue. Distrobox can export any binary a container can see.

**Comment by sekalengrengginang** [comment]
Integration with game engine like Unity,Godot,and maybe Unreal
Vscode pre-installed will be huge plus and maybe Docker too?

**Comment by dnkmmr69420** [comment]
> Integration with game engine like Unity,Godot,and maybe Unreal Vscode pre-installed will be huge plus and maybe Docker too?

what advantages does docker have over podman?

**Comment by badlydrawnface** [comment]
1. its requires root
2. its has a daemon service

not quite sure if those are really "advantages"

**Comment by sekalengrengginang** [comment]
more compatible with stuff
for example i tried to use devcontainers on bluefin image and it just PITA, meanwhile docker works ootb

**Comment by hisergiorojas** [comment]
I would like a distro that has preinstall the following software:

1. Blender
2. Godot
3. Visual Studio Code
4. Chrome
5. VFX Library https://vfxplatform.com/
6. Raven https://github.com/OpenTimelineIO/raven

**Comment by noelmiller** [comment]
Software I'd like to see that I haven't seen others post:
1. Virtual Machine Manager (layered seems to be the easiest way I've gotten it installed)
2. Podman Desktop
3. OBS

**Comment by AlexandrosMelissas** [comment]
I would suggest to have all the necessary stuff that are common for the majority of developers and are commonly used (podman, vscode, configs etc). As for the rest, they can be installed seperately.

**Comment by TaylorMichaelHall** [comment]
> > Integration with game engine like Unity,Godot,and maybe Unreal Vscode pre-installed will be huge plus and maybe Docker too?
> 
> what advantages does docker have over podman?

For me it's just a requirement for collaborating with others who all use Docker.

I would definitely appreciate an easy way to get docker up in Bazzite. It works in Bluefin out of the box, but getting docker working properly is a pain on just about every other atomic system that I've used.

**Comment by bayazidbh** [comment]
IMHO VMware, VirtualBox, and anything that requires kernel modules should be built in. At least the kernel module part - so that when user needs those tool, they can just be installed from a menu/command.

As long as that's done, I think everything else can just be thrown to different yafti menu/sub-categories.

**Comment by wojpawlik** [comment]
Why not start with only whatever -dx adds to Bluefin? Related: https://github.com/ublue-os/bluefin/issues/1321

**Comment by EPOCHvoyager** [comment]
### Some software suggestions as of yet:
- [_Reaper_](https://www.reaper.fm/) - **DAW for sound design and music production;** although technically paid software(some may call it 'nagware', in the vein of WinRAR), there's a 30-day trial, and it is distributed in the Arch `Extra` repository.
- [_Godots_](https://github.com/MakovWait/godots) - **Godot Engine version and project manager;** similar to the engine's standard project manager menu, but manages engine binaries as well, allows you to launch and assign specific projects with and to specific Godot versions, and download them to your machine - all-in-one.
- [_Furnace_](https://github.com/tildearrow/furnace) - **multi-system chiptune tracker;** simulates most known video game synthesizers - even _WonderSwan_ - within a single familiar interface.

**Comment by kittenvr** [comment]
Nodejs

**Comment by kittenvr** [comment]
Java gui

**Comment by p5** [comment]
There certainly needs to be a scope that needs to be referred to.

There's many suggestions of general development tools, rather than what I thought this was intended to be - a game dev image.

If we add everything from Bluefin DX, then we've just replicated Bluefin.

**Comment by SNThrailkill** [comment]
> There certainly needs to be a scope that needs to be referred to.
> 
> There's many suggestions of general development tools, rather than what I thought this was intended to be - a game dev image.
> 
> If we add everything from Bluefin DX, then we've just replicated Bluefin.

Would that really be the case? Because Bazzite has a lot of things I wanted for Bluefin DX like OpenRazer, Emudeck, OpenTabletDriver, Sunshine, Waydroid, etc. Whereas Bluefin has things I want like Jetbrains toolbox. cockpit, Gnome VRR, VFIO, ollama, etc. 

Im thinking about making my own image or configuration and I dont even know which would be the correct base because both seem to have what I want.

**Comment by VeachTech** [comment]
I want bluefin-dx but on bazzite. To me that would be the main base for gamedev. I wouldn't include any software that doesn't require configuration in order to fully function.
At a minimum:
I need devcontainers working out of the box with vs code.

**Comment by kittenvr** [comment]
I want java and nodejs

**Comment by sekalengrengginang** [comment]
> I want java and nodejs

for java and nodejs you can use homebrew

**Comment by kittenvr** [comment]
> > I want java and nodejs
> 
> for java and nodejs you can use homebrew

Java isn't on homebrew

**Comment by sekalengrengginang** [comment]
> > > I want java and nodejs
> > 
> > 
> > for java and nodejs you can use homebrew
> 
> Java isn't on homebrew

[`openjdk`](https://formulae.brew.sh/formula/openjdk)

**Comment by Confusedsiren56** [comment]
1 Tools for making Minecraft mods 
2 unreal engine 4 and 5 
3 VS code
4 Java
5 a version of bazzite that's lighter wight

**Comment by m2Giles** [comment]
My bad take.

Basically everything that's been mentioned can be run via containers/flatpak/brew. While having your editor on the host is nice, it doesn't actually need to be there. For example docker, incus, and libvirt will all happily run inside of a rootful container. Distrobox makes it even easier. It sounds more like people just want an easy button for configuring an option not necessarily it to be shipped on image.

What is the scope of -dx on Bazzite?

**Comment by KyleGospo** [comment]
> 1 Tools for making Minecraft mods 
> 2 unreal engine 4 and 5 
> 3 VS code
> 4 Java
> 5 a version of bazzite that's lighter wight

The current version is as light as it will get from a software standpoint.

DX will naturally be heavier.

**Comment by KyleGospo** [comment]
> My bad take.
> 
> Basically everything that's been mentioned can be run via containers/flatpak/brew. While having your editor on the host is nice, it doesn't actually need to be there. For example docker, incus, and libvirt will all happily run inside of a rootful container. Distrobox makes it even easier. It sounds more like people just want an easy button for configuring an option not necessarily it to be shipped on image.
> 
> What is the scope of -dx on Bazzite?

Intended scope of a bazzite-dx is inherently different from bluefin-dx. Our focus would be to have a game developer workstation that's ready to use from install.

**Comment by bi0shacker001** [comment]
> > My bad take.
> > Basically everything that's been mentioned can be run via containers/flatpak/brew. While having your editor on the host is nice, it doesn't actually need to be there. For example docker, incus, and libvirt will all happily run inside of a rootful container. Distrobox makes it even easier. It sounds more like people just want an easy button for configuring an option not necessarily it to be shipped on image.
> > What is the scope of -dx on Bazzite?
> 
> Intended scope of a bazzite-dx is inherently different from bluefin-dx. Our focus would be to have a game developer workstation that's ready to use from install.

wouldn't the logical approach to this simply be bundling common utilities with as many game engines as possible, so pretty much anyone can just boot it up and start working with their engine of choice?

**Comment by KyleGospo** [comment]
That's correct, but we can't appeal to everybody. That's why this is open for comment, we need to identify what the common tools are and what the uncommon tools are, make sure the common ones are installed where it makes sense, and everything else is easy to install when needed.

**Comment by RoyAwesome** [comment]
So, I do a lot of noodling on my spare time.  Professionally I'm a senior gameplay engineer, so most of the things I reach for are C++ related when I want to do gamedev noodling.  

In trying to get my at-home development workflow set up, I've been bouncing into quite a few issues that I think a DX image can solve.  

VS Code on flatpak is a nightmare to work with.  Putting into distrobox runs into issues doing gamedev with running 3d graphics.  

I would really like to see VS Code, CMake (and the vscode cmake extension), and llvm/clang working together out of the box.  Together they form the backbone of my development workflow, and getting them working in bazzite normally is kind of janky.  

Outside of that, I also noodle a lot with Godot, and with .net workflows.  Installing .net is kind of pain, and having that come pre-installed would be nice.

**Comment by wojpawlik** [comment]
Why doesn't [`brew install cmake`](https://formulae.brew.sh/formula/cmake) work for you? I'd be reluctant to add anything already available on brew or Flathub.

**Comment by m2Giles** [comment]
> Why doesn't [`brew install cmake`](https://formulae.brew.sh/formula/cmake) work for you? I'd be reluctant to add anything already available on brew or Flathub.

This is what I do. Vscode picks it up no problem with the cmake extension for me. I may have added homebrew to my argv.json PATH.

**Comment by RoyAwesome** [comment]
> Why doesn't [`brew install cmake`](https://formulae.brew.sh/formula/cmake) work for you? I'd be reluctant to add anything already available on brew or Flathub.

It worked okay, but a lot of features on it didn't work.  cmake wasn't able to find anything, i had to set everything up myself.  lldbg still doesn't work for me.  This would be nice for the -dx image to just have available or set up for me without having to google and hope i find something relevant for bazzite's setup.

**Comment by Ran-Mewo** [comment]
Suggestion:
- `rpm-ostree install openssl-devel -- --allow-replacement`

I'd the packages this command installs to be a thing by default so I don't have a pain trying to figure out why rust isn't working

**Comment by Rerum02** [comment]
I would like Zed on there, there is a Flatpak, but as of now, Nvidia on Wayland will not work, so Zed rpm would be nice

**Comment by bb010g** [comment]
I'd prefer a pairing of `bazzite-dx` and `bazzite-gamedev-dx`, where `bazzite{,-gnome}-dx` closely follows `{aurora,bluefin}-dx`, and `bazzite{,-gnome}-gamedev-dx` builds from that to include tools specific to gamedev. I understand wanting to give people an easy way to start game development on Linux, but my use case for `bazzite-dx` would be just doing some general development on the same laptop I play games with. I could use `aurora-dx` for that, but I'd like the gaming features in Bazzite. 

Additionally, building `bazzite-gamedev-dx` on top of `bazzite-dx` should make it both easier to keep in sync with `aurora-dx`, make it more apparent to those curious what tools are added for gamedev, and even allow `bazzite-dx` to ship while `bazzite-gamedev-dx` is still being planned.

@Sparkrai has been building unofficial `bazzite-dx` images using this approach at <https://github.com/Sparkrai/bazzite-dx>.

**Comment by mcreekmore** [comment]
> > I'd prefer a pairing of `bazzite-dx` and `bazzite-gamedev-dx`, where `bazzite{,-gnome}-dx` closely follows `{aurora,bluefin}-dx`, and `bazzite{,-gnome}-gamedev-dx` builds from that to include tools specific to gamedev.
> 
> This. Perhaps the installation of dev tools could be implemented in ujust like `ujust setup-dx-edition` or for each tool seperately (also mentioned in a previous reply). It would be more universal.

I would just like to +1 this approach. Without it, this would introduce many image variants: `bazzite * nvidia * dx * desktopenv * etc` in the future. Not sure how the community feels about that.

Currently Sparkrai's `bazzite-dx` image is interesting but doesn't include the nvidia images as of yet. When I find some time, I'd like to see if I could contribute to this effort.

**Comment by RoyAwesome** [comment]
vulkan debug layers would be really nice on a bazzite-dx

**Comment by bb010g** [comment]
Still relevant.

**Comment by castrojo** [comment]
Ok since we seem stuck here, look what I found!

https://artifacthub.io/packages/container/amyos/amyos

I know there are people on this issue that this would appeal to. It'd be a great way to get user feedback and then get people interested in prototyping more with our image templates. If you have any feedback from using this image please report your findings to the group!

**Comment by castrojo** [comment]
Ok I have an idea. Things are much simpler with the template, it looks relatively straightforward to just cut out what we don't need to solve this problem. 

I say we just make a new repo using this repo as a template, we cut all the stuff that's opinion to make a good base image that matches Bluefin and Aurora. Since it's in a separate repo we can build it after separate from Bazzite so it doesn't blow up the build matrix.

That's `bazzite-dx`. That covers the people who want to game and do dev.

Then we make a new repo called `bazzite-gdx`, again, based off our own template but `FROM bazzite-dx` and that becomes the game dev images. All the people who want godot, etc. can just roll that way. 

I think this solves this problem in a relatively quick amount of time, and as a bonus is something way more people can do, the core team can concentrate on Bazzite itself. This is just chaining two templates together. This is also an area where Bazzite can _really_ show value to a community vital to our success. GDX can go hard and DX keeps it open for people to slurp in what they like about GDX into their custom images.

Also we have the learnings from making Bluefin LTS and things are much simpler now. When we split Aurora out from Bluefin it gave people a new place to dive in, which is probably the smartest thing we can do right now, we need more folks diving in.

**Comment by castrojo** [comment]
Also the bluefin LTS repo has local-first justfiles for development, it would make local development of bazzite ready to rock and roll almost overnight.

**Comment by wolfyreload** [comment]
@castrojo I think that you are onto something there. I was also wondering how crazy the build matrix would get with adding dx images to the mix. That would solve everything imo. 

Would it make sense to go one step further and split the builds into builds by DE and maybe share a common bazzite-base image? It would make creating a new Bazzite DE such as support for Budgie or Cosmic much simpler when it comes out. And it would allow others to quite easily add support for the DE of their choice because they don't need to fork Bazzite but rather just build on top of Bazzite-Base.

**Comment by castrojo** [comment]
> Would it make sense to go one step further and split the builds into builds by DE

I reread this thread before starting this. 😄 

Bottom up and spider from there, this issue is from 2023 because of scope creep and "let's add more desktops". Here's the a core base image, testing and PRs accepted.

https://github.com/ublue-os/bazzite-dx

**Comment by castrojo** [comment]
Ok closing this so people can focus on their respective use cases.
